### PR TITLE
AOTI Minifier

### PR DESCRIPTION
**Pull request by yushangdi**
See documentation at https://docs-preview.pytorch.org/pytorch/pytorch/139351/torch.compiler_aot_inductor_minifier.html. 

Add a minifier for AOTI. 

Test Plan:
python test/inductor/test_minifier.py 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_dynamo/repro/after_aot.py [hunk 1]
```diff
@@ -14,7 +14,6 @@
 from importlib import import_module
 from tempfile import TemporaryFile
 from typing import Any, Callable, Dict, Sequence, TYPE_CHECKING, Union
-from typing_extensions import Unpack
```
Comment (yushangdi) [inline torch/_dynamo/repro/after_aot.py:17]: linter added change

File: docs/source/torch.compiler_aot_inductor_minifier.rst [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    # nvcc: NVIDIA (R) Cuda compiler driver
+    # Copyright (c) 2005-2023 NVIDIA Corporation
+    # Built on Fri_Jan__6_16:45:21_PST_2023
+    # Cuda compilation tools, release 12.0, V12.0.140
+    # Build cuda_12.0.r12.0/compiler.32267302_0
+
+    # GPU Hardware Info:
+    # NVIDIA PG509-210 : 8
+
+    exported_program = torch.export.load('/data/users/shangdiy/pytorch/torch_compile_debug/run_2024_10_31_16_48_02_720863-pid_3598491/minifier/checkpoints/exported_program.pt2')
```
Comment (desertfire) [inline docs/source/torch.compiler_aot_inductor_minifier.rst:191]: I expect the repro file to contain a trimmed model code which runs the full export-compile-run flow and then reproduces the problem. Loading another .pt2 file is not intuitive here.

File: torch/_inductor/__init__.py [hunk 3]
```diff
@@ -100,9 +131,52 @@ def aoti_compile_and_package(
 
     res = package_aoti(package_path, aoti_files)
     assert res == package_path
+
+    if load_and_run:
+        compiled_model = aoti_load_package(package_path)
+        aoti_result = compiled_model(*args)
```
Comment (desertfire) [inline torch/_inductor/__init__.py:137]: We could compare with eager result here and thus enable accuracy minifier. Ok to do it in a follow-up PR.

Comment (yushangdi) [reply torch/_inductor/__init__.py:137]: > We could compare with eager result here and thus enable accuracy minifier. Ok to do it in a follow-up PR.

yeah, I can do it in a follow up

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -42,6 +42,9 @@ def bundled_autotune_remote_cache_default() -> Optional[bool]:
 # Whether to enable printing the source code for each future
 verbose_progress = False
 
+# dump an aoti minifier if program errors
+dump_aoti_minifier: bool = os.environ.get("DUMP_AOTI_MINIFIER", "0") == "1"
```
Comment (muchulee8) [inline torch/_inductor/config.py:46]: I would recommend moving this under `aot_inductor`.

**Review by desertfire — APPROVED** [review]

**Comment by henrylhtsang** [comment]
Thanks for the work. Not sure how much work that would be, but is it possible to print some metadata for the input arguments, so users can try to reproduce them without needing to load local files?

e.g. for tensor, print shape and dtype. For python constants, print everything.

**Comment by yushangdi** [comment]
> Thanks for the work. Not sure how much work that would be, but is it possible to print some metadata for the input arguments, so users can try to reproduce them without needing to load local files?
> 
> e.g. for tensor, print shape and dtype. For python constants, print everything.

Thanks for the suggestion! Yeah I can add that, it shouldn't be too hard.
